sys/include/net/gcoap.h

```diff
@@ -48,7 +48,7 @@
  * this by uncommenting the appropriate lines in gcoap's make file.
  *
  * gcoap allows an application to specify a collection of request resource paths
- * it wants to be notified about. Create an array of resources (coap_resource_t
+ * it wants to be notified about. Create an array of resources (gcoap_resource_t
```

Comment:
the same might be done for the link_encoder

(i like these changes as they clearup where a structure is defined)

Reply:
Unfortunately this causes some issues with the Rust wrappers which in turn require a CI upgrade :(

So if you don't insist on splitting this, could we move forward with this as a single PR?

Reply:
wouldn't this PR trigger the same issue (if we restart murdock now) until rust is upgraded

Reply:
`gcoap.h:210-308` explains how to do block transfers with gcoap:
the nanocoap functions are used  i think they are missing the initialization for the respose structure after this PR is mergerd.

This section also provides a link to an example.

Reply:
> wouldn't this PR trigger the same issue (if we restart murdock now) until rust is upgraded

no, the rust wrapper uses the presence of `coap_build_pkt_t` to detect whether to use the old or the new API.
It has now been fixed to check for `gcoap_resource_t` instead.

> the nanocoap functions are used i think they are missing the initialization for the respose structure after this PR is mergerd.

Maybe I miss something, but those are not using any of the functions that makes use of `coap_build_pkt_t`?
The new type is not used by gcoap (that's why I introduced `gcoap_resource_t`) - it should be, but I was hoping this could be done as a follow-up.

Reply:
gcoap is using the nanocoap functions to generate the header coap_pkgs, while most of these generator functions do not make any use of `buf and len`/`coap_build_pkt_t`, block transfer does ->  with this pr gcoap would loose nanocoaps block-transfer abilities (which is described in gcoaps documentation (also featuring an external example).

I am no sure about the state of #17168 especial in the looking at block-tranfers. May be this API change is also welcome to gcoap.

While this PR tries to break just nanocoaps-API it implicitly also breaks gcoaps-API (@block-transfers).

maybe @chrysn or @miri64 have some insight into what changes are going to happen in gcoap maybe this API-change is welcome and should just be forwarded to gcoap ( which might be easier since it may not need the introduction of 'gcoap_rescource_t').

Reply:
#17801 implements caching for gcoap and somehow also for nanocoap (using the old buf an len but introducing a nanocoap_-prefix)

Reply:
> while most of these generator functions do not make any use of `buf and len`/`coap_build_pkt_t`, block transfer does -> with this pr gcoap would loose nanocoaps block-transfer abilities (which is described in gcoaps documentation (gcoap.h:210-308)

Please explain to me which function(s) you mean. If Gcoap would use on of the converted nanocoap functions, that would be a compilation error as Gcoap does not yet make use of `coap_build_pkt_t`.

There are functions in `nanocoap.c` that are exclusively used by Gcoap, but as far as I can tell those are not touched by this PR.